torch/nn/functional.py

```diff
@@ -3712,9 +3727,34 @@ def huber_loss(
         )
 
     expanded_input, expanded_target = torch.broadcast_tensors(input, target)
-    return torch._C._nn.huber_loss(
-        expanded_input, expanded_target, _Reduction.get_enum(reduction), delta
-    )
+
+    if weights is None:
+        # Use the optimized C++ backend for standard Huber loss
+        return torch._C._nn.huber_loss(
+            expanded_input, expanded_target, _Reduction.get_enum(reduction), delta
+        )
+    else:
+        if weights.size() != input.size():
+            raise ValueError("Weights and input must have the same size.")
+
+        # Calculate the unweighted loss first
+        unweighted_loss = torch._C._nn.huber_loss(
+            expanded_input, expanded_target, _Reduction.get_enum("none"), delta
+        )
+
+        # Apply weights to the unweighted loss
+        weighted_loss = unweighted_loss * weights
+
+        if reduction == "none":
+            return weighted_loss
+        elif reduction == "sum":
+            return torch.sum(weighted_loss)
+        elif reduction == "mean":
+            return torch.sum(weighted_loss) / torch.sum(weights)
```

Comment:
hm, why is this not `weighted_loss.mean()`? which seems to be the precedent for bce_loss with weights as well as multimarginloss with weights in this file, similarly below https://github.com/pytorch/pytorch/blob/46935c8241db2df5ea2d5da6025d53b343f3e094/torch/nn/functional.py#L3958-L3959